### DISCUSSION TITLE
Is it possible to update data after mutation without using useQuery hook higher up in the tree?

### DISCUSSION THREAD

**Discussion by PiotrFidurski** [TanStack/query#2379] [Q&A] [answered]
Hello I know that the title might be confusing/misleading, so I'll try to explain. I have a simple component that gets an array of data from the server, maps to post components and each of them can be liked/disliked by users. My question is, can I do something like: 

https://codesandbox.io/s/hardcore-morning-jgi4h?file=/components/PostContainer.js

```javascript
function Home({
  posts,
}: 
InferGetServerSidePropsType<typeof getServerSideProps>) {

  return (
    <>
      <div
        css={{
          display: "flex",
          flexGrow: 1,
          justifyContent: "space-between",
          position: "relative",
          alignItems: "flex-start",
          [mQ("1000")]: { flexDirection: "column", alignItems: "center" },
        }}
      >
        <>
          <Feed posts={posts} />
        </>
      </div>
    </>
  );
}

export const getServerSideProps: GetServerSideProps = async (context) => {
  const queryClient = new QueryClient();
  await queryClient.prefetchQuery("posts", async () => {
    const postsResponse = await fetch("http://localhost:3000/api/posts", {
      method: "GET",
    });
    return await postsResponse.json();
  });

  const withKeys: any = [];
  dehydrate(queryClient).queries.forEach(
    (query) =>
      (withKeys[query.queryKey as any] = (query.state.data as any).data)
  );

  return {
    props: {
      posts: withKeys["posts"],
      // sending data directly instead of dehydratedState: dehydrate(queryClient),
      users: users.data,
    },
  };
};
```
The data is available and all that, but when using useMutation with onSuccess hook the data doesnt get updated, only after refresh/location change.

```javascript
  const mutation = useMutation(() => likePost(post._id), {
    onSuccess: () => {
      queryClient.invalidateQueries("posts");
    },
  });
```
I'm assuming its because we only prefetched query in ```getServerSideProps``` and never used ```useQuery``` hook from within Home component, is it possible to mutate data without using ```useQuery``` hook higher up in the tree?

**Comment by TkDodo** [comment, marked as answer]
Are you hydrating the queryClient on the frontend?

**Reply by PiotrFidurski** [reply]
Yes here's what my code looks like:

```javascript
function MyApp({ Component, pageProps }: AppProps) {
  const [queryClient] = React.useState(() => new QueryClient());

  return (
    <Provider session={pageProps.session}>
      <QueryClientProvider client={queryClient}>
        <Hydrate state={pageProps.dehydratedState}>
            <MainContainer>
              <Component {...pageProps} />
            </MainContainer>
        </Hydrate>
      </QueryClientProvider>
    </Provider>
  );
}
```

the first ```Provider``` in the tree comes from next-auth

**Reply by TkDodo** [reply]
well I'm seeing in your backend code:

```
      // sending data directly instead of dehydratedState: dehydrate(queryClient),
      users: users.data,
```

so if you are not sending down a `dehydratedState`, how can this work:

```
<Hydrate state={pageProps.dehydratedState}>
```

**Reply by PiotrFidurski** [reply]
Oh, I'm sending ```dehydratedState: dehydrate(queryClient)``` in my actual code, I commented it out in the example because I thought it'd help to explain what I'm trying to do, my actual code looks pretty much like what the docs say, and everything works perfectly when I add ```useQuery``` hook to ```Home``` component, I even changed from ```queryClient.invalidateQueries("posts")``` to ```setQueryData``` to not waste network call.

```javascript
queryClient.setQueryData("posts", (posts: any) => {
        return {
          ...posts,
          data: posts.data.filter((p: any) => {
            if (p._id === post._id) {
              p.likes = data.data.likes;
              p.likesCount = data.data.likesCount;
              return p;
            }
            return p;
          }),
        };
      });
````

**Reply by TkDodo** [reply]
okay, then I don't really know. Can you put your code in a codesandbox?

**Reply by PiotrFidurski** [reply]
Sure, do note that I don't do any API calls but the behavior I'm talking about is there https://codesandbox.io/s/hardcore-morning-jgi4h?file=/components/PostContainer.js

**Reply by TkDodo** [reply]
thank you, but I don't understand the issue. It seems the rehydration works, and then you can work with the query cache. If you send props down manually and just access them via `props.posts`, they are not tied at all to the query cache and will thus not update when you update something _in_ the query cache...

**Reply by PiotrFidurski** [reply]
ah, that's what I thought, this question was more of a "what if" rather than an issue with something, Thanks for answering and your time. 👍
